### DISCUSSION TITLE
Error on initial render of custom hook with window object

### DISCUSSION THREAD

**Discussion by hadikman** [vercel/next.js#72897] [Help] [answered]
### Summary

Hi
In every custom hook that uses window/local storage/history/navigation object, you must check `window !== undefind` to prevent of showing error on first render (If the window is available, the following error will disappear):
![Screenshot 2024-11-17 135715](https://github.com/user-attachments/assets/07d196b4-094d-4331-b4fb-c9e5bdd9e21c)

### The problem:
Is there any way to remove the window object checking (I'd like to remove this redundant if)?

In the following custom hook I checked the window if it's available then listen on event and I wouldn't have the mentioned error:
``` javascript
export default function useScreenSize() {
  const { isWindowReady } = useWindowReady()
  const [screenSize, setScreenSize] = React.useState()

  React.useLayoutEffect(() => {
    const handleResize = () => {
      setScreenSize(getScreenSize())
    }
    if (isWindowReady) {
      if (!screenSize) setScreenSize(getScreenSize())
      window.addEventListener('resize', handleResize)
    }

    return () => {
      window.removeEventListener('resize', handleResize)
    }
  }, [isWindowReady, screenSize])

  return screenSize
}

function getScreenSize() {
  return window.innerWidth 
}
```

Thank you in advance

**Comment by icyJoseph** [comment, marked as answer]
`useLayoutEffect` and `useEffect` are safe places to use `window` and other document access. You don't need to check if the `window` is ready within them, nor do you need a support hook for that. 

In your case, I'd say you can use `useSyncExternalStore` though. It is specifically designed to sync with state that lives elsewhere, such as the window dimensions, when a subscriber exists:

- https://react.dev/reference/react/useSyncExternalStore#adding-support-for-server-rendering
- https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api#another-example-scrolly
